File: LeetCode/medium/19_remove-nth-node-from-end-of-list/java/Solution.java
/*
 * @Descripttion: 
 * @version: 
 * @Author: renjie.zhang
 * @Date: 2019-10-24 17:30:27
 * @LastEditTime: 2019-10-24 17:43:21
 */
class Solution{
    public static void main(String[] args) {
        ListNode node1 = new ListNode(1);
        ListNode node2 = new ListNode(2);
        ListNode node3 = new ListNode(3);
        ListNode node4 = new ListNode(4);
        ListNode node5 = new ListNode(5);
        node1.next = node2;
        node2.next = node3;
        node3.next = node4;
        node1.next = node5;
        ListNode node =  removeNthFromEnd(node1, 2);
        System.out.print(node.val);
    }

    public static ListNode removeNthFromEnd(ListNode head, int n) {
        int length = 1;
        ListNode temp = head;
        while (temp.next != null){
            length++;
            temp = temp.next;
        }
        if(n<0||n>length){
            return null;
        }
        if(n == length){
            head = head.next;
            return head;
        }
        
        ListNode cureent = head;
        for (int i= length-n-1;i>0;i--){
            cureent = cureent.next;
        }
        
        cureent.next = cureent.next.next;
        return head;
    }
}

class ListNode{
    int val;
    ListNode next;
    ListNode(int x){
        this.val = x;
    }
}

File: LeetCode/medium/19_remove-nth-node-from-end-of-list/golang/Solution.go
/*
 * @Descripttion:删除链表的倒数第n个结点
 * @version:
 * @Author: renjie.zhang
 * @Date: 2019-10-24 17:30:08
 * @LastEditTime: 2019-11-05 21:12:15
 */
package golang

type ListNode struct {
	Val  int
	Next *ListNode
}

/*
删除链表的倒数第n个结点
*/
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	length := 1
	temp := head
	for temp.Next != nil {
		length = length + 1
		temp = temp.Next
	}
	if n < 0 || n > length {
		return nil
	}
	if n == length {
		return head
	}

	current := head

	for i := length - n - 1; i > 0; i-- {
		current = current.Next
	}
	current.Next = current.Next.Next
	return head
}
